src/js/autocomplete.js

```diff
@@ -1,14 +1,247 @@
+	const loading = document.createElement('div');
+	loading.classList.add('o-autocomplete__menu-loading');
+	loadingContainer.appendChild(loading);
+	return loadingContainer;
+}
+
+/**
+ * @callback PopulateResults
+ * @param {Array<string>} results - The results to show in the suggestions dropdown
+ * @returns {void}
+ */
+
+/**
+ * @callback Suggestions
+ * @param {string} query - Text which was typed into the autocomplete by the user
+ * @param {PopulateResults} populateResults - Function to call when ready to update the suggestions dropdown
+ * @returns {void}
+*/
+
+
+/**
+ * @typedef {Object} AutocompleteOptions
+ * @property {Suggestions} source - The function which retrieves the suggestions to display
+ */
+
 class Autocomplete {
 	/**
 	 * Class constructor.
 	 * @param {HTMLElement} [autocompleteEl] - The component element in the DOM
-	 * @param {Object} [options={}] - An options object for configuring the component
+	 * @param {AutocompleteOptions} [options={}] - An options object for configuring the component
 	 */
 	constructor (autocompleteEl, options) {
 		this.autocompleteEl = autocompleteEl;
-		this.options = Object.assign({}, {
+		this.options = Object.assign({
+			placeholder: '',
+			cssNamespace: 'o-autocomplete',
+			displayMenu: 'overlay',
+			showNoOptionsFound: false,
+			templates: {
+				suggestion: this.suggestionTemplate.bind(this)
+			}
 		}, options || Autocomplete.getDataAttributes(autocompleteEl));
+
+		if (this.options.source) {
+			// If source is a string, then it is the name of a global function to use.
+			// If source is not a string, then it is a function to use.
+			/**
+			 * @type {Suggestions}
+			 */
+			const customSuggestions = typeof this.options.source === 'string' ? window[this.options.source] : this.options.source;
+			/**
+			 * @param {string} query - Text which was typed into the autocomplete by the user
+			 * @param {PopulateResults} populateResults - Function to call when ready to update the suggestions dropdown
+			 * @returns {void}
+			*/
+			this.options.source = (query, populateResults) => {
+				this.showLoadingPane();
+				/**
+				 * @param {Array<string>} results - The results to show in the suggestions dropdown
+				 * @returns {void}
+				 */
+				const callback = (results) => {
+					this.hideLoadingPane();
+					populateResults(results);
+				};
+				customSuggestions(query, callback);
+			};
+			const input = autocompleteEl.querySelector('input');
+			const id = input.getAttribute('id');
+			autocompleteEl.innerHTML = '';
+			const options = Object.assign({
+				element: autocompleteEl,
+				id: id,
+			}, this.options);
+			accessibleAutocomplete(options);
+		} else {
+			const element = autocompleteEl.querySelector('select');
+			autocompleteEl.appendChild(element);
+			const options = Object.assign({
+				selectElement: element,
+				defaultValue: '',
+			}, this.options);
+			options.autoselect = false; // TODO: Find out if we should allow autoselect/hinting in the input
```

Comment:
I forget what these features are. Should we? :D

Reply:
Reading this comment back it sounds enthusiastic for adding features when actually let's not if we don't need them yet 😅

Reply:
```suggestion
			options.autoselect = false;
```